OgreMain/include/OgreParticles.h

```diff
@@ -0,0 +1,90 @@
+/*
+-----------------------------------------------------------------------------
+This source file is part of OGRE
+(Object-oriented Graphics Rendering Engine)
+For the latest info, see http://www.ogre3d.org/
+
+Copyright (c) 2000-2014 Torus Knot Software Ltd
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+-----------------------------------------------------------------------------
+*/
+#ifndef __Particles2_H__
+#define __Particles2_H__
+
+#include "OgrePrerequisites.h"
+#include "OgreHeaderPrefix.h"
+#include "OgreColourValue.h"
+#include "OgreMath.h"
+#include "OgreVector.h"
+
+#include "sh/cpp/block_n.hpp"
+
+namespace Ogre {
+
+
+    struct ParticleTtl
+    {
+        Real remaining;
+        Real total;
+    };
+
+    struct ParticleTranslation
+    {
+        Vector3 position;
+        Vector3 direction;
+    };
+
+    struct ParticleRotation
+    {
+        Real angle;
+        Real speed;
+    };
+
+    struct ParticleSize
+    {
+        Real width;
+        Real height;
+    };
+
+    struct Particles2 : public FXAlloc, private block_n <ParticleTtl, ParticleTranslation, ParticleRotation, ColourValue, ParticleSize>
```

Comment:
This class is quite close to Ogre::VertexData...

Reply:
so you're saying move the data to a vertex buffer instead of cpu?

Reply:
If you use a shadow buffer, you can ensure that the data will be read/ writable on the CPU. This is a thing I am still figuring out though.
This Particles2 class is a nicer API for that shadow buffer in a sense. Also this interferes with the ParticleRenderer that currently owns vertex buffers and takes care of quad creation.
With sv/gl_vertexid one could do that on the GPU though..

Reply:
i'm still using directx 9 so i really need to get up to date on shareable buffers etc.  same with the sv/gl_vertexid concept.

Reply:
see e.g. https://webgl2fundamentals.org/webgl/lessons/webgl-drawing-without-data.html

Reply:
awesome, will take a look.  thanks!